src/twisted/internet/defer.py

```diff
@@ -726,140 +838,145 @@ def _runCallbacks(self):
         try:
             createFuture = loop.create_future
         except AttributeError:
-            from asyncio import Future
 
-            def createFuture():
+            def createFuture() -> Future:
                 return Future(loop=loop)
 
         future = createFuture()
 
-        def checkCancel(futureAgain):
+        def checkCancel(futureAgain: Future) -> None:
             if futureAgain.cancelled():
                 self.cancel()
 
-        def maybeFail(failure):
+        def maybeFail(failure: Failure) -> None:
             if not future.cancelled():
                 future.set_exception(failure.value)
 
-        def maybeSucceed(result):
+        def maybeSucceed(result: object) -> None:
             if not future.cancelled():
                 future.set_result(result)
 
         self.addCallbacks(maybeSucceed, maybeFail)
         future.add_done_callback(checkCancel)
+
         return future
 
     @classmethod
-    def fromFuture(cls, future):
+    def fromFuture(
+        cls, future: Future[_DeferredResultT]
+    ) -> "Deferred[_DeferredResultT]":
         """
-        Adapt an L{asyncio.Future} to a L{Deferred}.
+        Adapt an L{Future} to a L{Deferred}.
 
-        @note: This creates a L{Deferred} from a L{asyncio.Future}, I{not} from
+        @note: This creates a L{Deferred} from a L{Future}, I{not} from
             a C{coroutine}; in other words, you will need to call
-            L{asyncio.ensure_future},
-            L{asyncio.loop.create_task} or create an
+            L{asyncio.ensure_future}, L{asyncio.loop.create_task} or create an
             L{asyncio.Task} yourself to get from a C{coroutine} to a
-            L{asyncio.Future} if what you have is an awaitable coroutine and
-            not a L{asyncio.Future}.  (The length of this list of techniques is
+            L{Future} if what you have is an awaitable coroutine and
+            not a L{Future}.  (The length of this list of techniques is
             exactly why we have left it to the caller!)
 
         @since: Twisted 17.5.0
 
-        @param future: The Future to adapt.
-        @type future: L{asyncio.Future}
+        @param future: The L{Future} to adapt.
 
-        @return: A Deferred which will fire when the Future fires.
-        @rtype: L{Deferred}
+        @return: A L{Deferred} which will fire when the L{Future} fires.
         """
 
-        def adapt(result):
-            try:
-                extracted = result.result()
-            except BaseException:
-                extracted = failure.Failure()
-            adapt.actual.callback(extracted)
+        class Adaptor:
+            def __init__(self, actual: Deferred[Any]) -> None:
+                self.actual = actual
+
+            def __call__(self, result: Future) -> None:
+                try:
+                    extracted = result.result()
+                except BaseException:
+                    extracted = Failure()
+                self.actual.callback(extracted)
```

Comment:
This to avoid a function that sets an attribute on itself

Reply:
It might be less disruptive to use `nonlocal`; the only reason this used a function attribute was to avoid py3-only features.

Reply:
OK replaced with use of `nonlocal`

Reply:
New diff only adds type hints, no code changes other than to make `actual` its own variable (marked `nonlocal` in `uncancel`, where it is set), instead of an attribute on the `adapt` function.